go/service/chat.go

```diff
@@ -4,38 +4,301 @@
+	rarg := chat1.GetThreadRemoteArg{
+		ConversationID: arg.ConversationID,
+		Pagination:     arg.Pagination,
+	}
+	boxed, err := h.remoteClient().GetThreadRemote(ctx, rarg)
+	if err != nil {
+		return keybase1.ThreadView{}, err
+	}
+
+	return h.unboxThread(ctx, boxed)
 }
 
-func (h *chatLocalHandler) NewConversationLocal(context.Context, chat1.ConversationIDTriple) error {
-	return nil
+// NewConversationLocal implements keybase.chatLocal.newConversationLocal protocol.
+func (h *chatLocalHandler) NewConversationLocal(ctx context.Context, trip chat1.ConversationIDTriple) error {
+	md := chat1.ConversationMetadata{
+		IdTriple: trip,
+	}
+	return h.remoteClient().NewConversationRemote(ctx, md)
+}
+
+// PostLocal implements keybase.chatLocal.postLocal protocol.
+func (h *chatLocalHandler) PostLocal(ctx context.Context, arg keybase1.PostLocalArg) error {
+	// encrypt the message
+	boxed, err := h.boxMessage(ctx, arg.MessagePlaintext)
+	if err != nil {
+		return err
+	}
+
+	// get device signing key for this user
+	signingKey, err := engine.GetMySecretKey(h.G(), h.getSecretUI, libkb.DeviceSigningKeyType, "sign chat message")
+	if err != nil {
+		return err
+	}
+	kp, ok := signingKey.(libkb.NaclSigningKeyPair)
+	if !ok || kp.Private == nil {
+		return libkb.KeyCannotSignError{}
+	}
+
+	// sign the header, encrypted body
+	if err := h.signMessageBoxed(&boxed, kp); err != nil {
+		return err
+	}
+
+	// post to remote gregord
+	rarg := chat1.PostRemoteArg{
+		ConversationID: arg.ConversationID,
+		MessageBoxed:   boxed,
+	}
+
+	return h.remoteClient().PostRemote(ctx, rarg)
+}
+
+// getSecretUI returns a SecretUI, preferring a delegated SecretUI if
+// possible.
+func (h *chatLocalHandler) getSecretUI() libkb.SecretUI {
+	ui, err := h.G().UIRouter.GetSecretUI(0)
+	if err == nil && ui != nil {
+		h.G().Log.Debug("chatLocalHandler: using delegated SecretUI")
+		return ui
+	}
+	h.G().Log.Debug("chatLocalHandler: using local SecretUI")
+	return h.BaseHandler.getSecretUI(0, h.G())
+}
+
+// remoteClient returns a client connection to gregord.
+func (h *chatLocalHandler) remoteClient() *chat1.RemoteClient {
+	return &chat1.RemoteClient{Cli: h.gh.cli}
+}
+
+// unboxThread transforms a chat1.ThreadViewBoxed to a keybase1.ThreadView.
+func (h *chatLocalHandler) unboxThread(ctx context.Context, boxed chat1.ThreadViewBoxed) (keybase1.ThreadView, error) {
+	thread := keybase1.ThreadView{
+		Pagination: boxed.Pagination,
+	}
+
+	finder := newKeyFinder()
+	for _, msg := range boxed.Messages {
+		unboxed, err := h.unboxMessage(ctx, finder, msg)
+		if err != nil {
+			return keybase1.ThreadView{}, err
+		}
+		thread.Messages = append(thread.Messages, unboxed)
+	}
+
+	return thread, nil
+}
+
+// unboxMessage unboxes a chat1.MessageBoxed into a keybase1.Message.  It finds
+// the appropriate keybase1.CryptKey.
+func (h *chatLocalHandler) unboxMessage(ctx context.Context, finder *keyFinder, msg chat1.MessageBoxed) (keybase1.Message, error) {
```

Comment:
Would you mind having a `type boxer` that has these (un)boxMessage methods? This could be used by https://github.com/keybase/client/pull/3859 as well. If not, I can move them in that PR after this is merged.